### DISCUSSION TITLE
Enable use of standby Postgresql nodes for RO queries

### DISCUSSION THREAD

**Discussion by udf2457** [directus/directus#15140] [Feature Requests]
As you are no doubt aware, Postgres can be configured in a mode with primary and one or more replicas.

Other than the obvious high-availability and replication benefits, the other key benefit is that the standby server(s) can be used for read-only queries by clients, the so-called [Hot-Standby mode](https://www.postgresql.org/docs/14/hot-standby.html).

It would make a lot of sense if Directus could be configured such that it could send all / a proportion / round-robin read-only queries to standby server(s).   

For busy database systems which may be serving more parts of the business than just Directus, this would mean that un-necessary queries against the primary server could be reduced.

For example, a website user Directus role whose permissions are 100% read only (i.e. only retrieving display data for the pages).  That could be dealt with entirely by a standby Postgres database because by definition all queries would be read-only queries.

**Comment by rijkvanzanten** [comment]
We do have to keep in mind that we shouldn't use the read replica inside of transactions where items are written, as those reads might need the data that only exists in the transactions 🤔

**Comment by udf2457** [comment]
@rijkvanzanten That's true, although I would classify a "transaction" the same as a single `INSERT`, `UPDATE` or `DELETE`.  So yes of course that would need to be done against the primary server.

But also with Postgres, maybe you should consider migrating externally managed transactions to Postgres stored functions anyway.

**Comment by udf2457** [comment]
Also if concurrency related race conditions are a concern then it might also be worth considering optimistic locking.

Postgres doesn't have a built-in equivalent to, say, Oracle's `ora_rowscn`, so you'd have to add the column yourself (or use timestamp columns), but basically it works like this:

```
select foo,bar,change_number from foobar where x=y;
update foobar set foo=oof,change_number=change_number+1 where x=y and change_number=old_change_number;

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
